File: crates/python/mirascript/_vm/operations/iterable.py
from __future__ import annotations
from typing_extensions import Iterable as IterableType

from ..._helpers.convert import to_string
from ..._helpers.types import is_vm_array, is_vm_primitive, is_vm_record, is_vm_wrapper
from ..types import VmAny, VmValue
from ..error import VmError
from .common import AssertInit
from .type_check import Type
from .utils import is_same


def In(value: VmAny, iterable: VmAny) -> bool:
    AssertInit(value)
    if is_vm_array(iterable):
        if value is None:
            return value in iterable

        for item in iterable:
            if is_same(value, item):
                return True

        return False
    if is_vm_primitive(iterable):
        return False
    pk = to_string(value)
    if is_vm_record(iterable):
        return pk in iterable
    if is_vm_wrapper(iterable):
        return iterable.has(pk)

    AssertInit(iterable)
    return False


def Iterable(value: VmAny) -> IterableType[VmValue]:
    AssertInit(value)
    if is_vm_wrapper(value):
        return value.keys()
    if is_vm_array(value):
        return value
    if is_vm_record(value):
        return value.keys()
    raise VmError(f"`Value is not iterable {Type(value)}", None)

File: packages/mirascript/src/vm/operations/iterable.ts
import { VmError } from '../../helpers/error.js';
import { hasOwnEnumerable, keys } from '../../helpers/utils.js';
import { toString } from '../../helpers/convert/index.js';
import { display } from '../../helpers/serialize.js';
import { isVmPrimitive, isVmArray, isVmFunction, isVmWrapper } from '../../helpers/types/index.js';
import type { VmAny, VmRecord, VmValue } from '../types/index.js';
import { isSame } from './utils.js';
import { $AssertInit } from './common.js';

/** 检查值是否在可迭代对象中 */
export const $In = (value: VmAny, iterable: VmAny): boolean => {
    $AssertInit(value);
    $AssertInit(iterable);
    if (iterable == null) return false;
    if (typeof iterable != 'object') return false;
    if (isVmArray(iterable)) {
        if (value == null) {
            // array may have empty slots
            for (const item of iterable) if (item == null) return true;
            return false;
        }
        // JS %SameValueZero is same with `isSame` in this context
        if (isVmPrimitive(value)) return iterable.includes(value);
        // value is not null here, so it's ok to skip empty slots, since `isSame(null, something)` is always false
        return iterable.some((item = null) => isSame(item, value satisfies NonNullable<VmValue>));
    }
    // iterable is a record or an extern here, value should be a string
    const key = toString(value, undefined);
    if (isVmWrapper(iterable)) return iterable.has(key);
    return hasOwnEnumerable(iterable satisfies VmRecord, key);
};

/** 获取可迭代对象 */
export const $Iterable = (value: VmAny): Iterable<VmValue | undefined> => {
    $AssertInit(value);
    if (isVmWrapper(value)) return value.keys();
    if (isVmArray(value)) return value;
    if (value != null && typeof value == 'object') return keys(value);
    throw new VmError(`Value is not iterable: ${display(value)}`, isVmFunction(value) ? [] : [value]);
};
